scheduler/crates/api/src/user/get_multiple_users_freebusy.rs

```diff
@@ -0,0 +1,279 @@
+            .await?;
+
+        Ok(GetMultipleFreeBusyResponse(busy_event_instances))
+    }
+}
+
+impl GetMultipleFreeBusyUseCase {
+    async fn get_calendars_from_user_ids(&self, ctx: &NettuContext) -> Vec<Calendar> {
+        join_all(
+            self.user_ids
+                .iter()
+                .map(|user_id| ctx.repos.calendars.find_by_user(user_id)),
+        )
+        .await
+        .into_iter()
+        .flatten()
+        .collect()
+    }
+
+    async fn get_event_instances_from_calendars(
+        &self,
+        timespan: &TimeSpan,
+        ctx: &NettuContext,
+        calendars: Vec<Calendar>,
+    ) -> Result<HashMap<ID, VecDeque<EventInstance>>, UseCaseError> {
+        // For quick lookup by calendar id
+        let calendars_lookup = calendars
+            .iter()
+            .map(|cal| (cal.id.to_string(), cal))
+            .collect::<HashMap<_, _>>();
+
+        // End result
+        let mut events_per_user = HashMap::new();
+
+        // Fetch all events for all calendars
+        // This is not executed yet (lazy)
+        let all_events_futures = calendars.iter().map(|calendar| async move {
+            let events = ctx
+                .repos
+                .events
+                .find_by_calendar(&calendar.id, Some(timespan))
+                .await
+                .unwrap_or_default(); // TODO: Handle error
+            Ok((calendar.user_id.clone(), events)) as Result<(ID, Vec<CalendarEvent>), UseCaseError>
+        });
+
+        // Fetch events in chunks of 5
+        let mut futures_stream = stream::iter(all_events_futures).chunks(5);
+
+        // Fetch events in parallel (actual execution)
+        while let Some(futures) = futures_stream.next().await {
+            let events_res = join_all(futures).await;
+
+            for event_result in events_res {
+                match event_result {
+                    Ok((user_id, events)) => {
+                        let expanded_events =
+                            self.expand_events(events, timespan, &calendars_lookup);
+                        events_per_user.insert(user_id, expanded_events);
+                    }
+                    Err(e) => return Err(e),
+                }
+            }
+        }
+
+        Ok(events_per_user)
+    }
+
+    fn expand_events(
+        &self,
+        events: Vec<CalendarEvent>,
+        timespan: &TimeSpan,
+        calendars_lookup: &HashMap<String, &Calendar>,
+    ) -> VecDeque<EventInstance> {
+        events
+            .into_iter()
+            .map(|event| {
+                let calendar = calendars_lookup
+                    .get(&event.calendar_id.to_string())
+                    .unwrap();
+                event.expand(Some(timespan), &calendar.settings)
+            })
+            .filter(|instances| !instances.is_empty())
+            .flat_map(|instances| CompatibleInstances::new(instances).inner())
+            .collect()
+    }
+}
+
+#[cfg(test)]
```

Comment:
Is it idiomatic to put the test in the same file in Rust?

Reply:
It depends:
- it's idiomatic to have unit tests in the same file
- it's idiomatic to have integration tests in a specific "tests" directory

It's worth noting that having unit tests in the same file allows to access private functions, structs, etc., of that module (as the tests are effectively inside the module)

([source](https://doc.rust-lang.org/book/ch11-03-test-organization.html))

Reply:
TIL! I like that actually.